CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc

```diff
@@ -0,0 +1,855 @@
+          clust_.minPixelCol = cluster->minPixelCol();
+          clust_.minPixelRow = cluster->minPixelRow();
+
+          // fill the trackhit info
+          TrajectoryStateOnSurface tsos = itTraj->updatedState();
+          if (!tsos.isValid()) {
+            edm::LogWarning("SiPixelLorentzAnglePCLWorker") << "tsos not valid" << std::endl;
+            continue;
+          }
+          LocalVector trackdirection = tsos.localDirection();
+          LocalPoint trackposition = tsos.localPosition();
+
+          if (trackdirection.z() == 0)
+            continue;
+          // the local position and direction
+          trackhit_.alpha = atan2(trackdirection.z(), trackdirection.x());
+          trackhit_.beta = atan2(trackdirection.z(), trackdirection.y());
+          trackhit_.gamma = atan2(trackdirection.x(), trackdirection.y());
+          trackhit_.x = trackposition.x();
+          trackhit_.y = trackposition.y();
+
+          // get qScale_ = templ.qscale() and  templ.r_qMeas_qTrue();
+
+          float cotalpha = trackdirection.x() / trackdirection.z();
+          float cotbeta = trackdirection.y() / trackdirection.z();
+          float cotbeta_min = clustSizeYMin_ * ypitch_ / width_;
+          if (fabs(cotbeta) <= cotbeta_min)
+            continue;
+          double drdz = sqrt(1. + cotalpha * cotalpha + cotbeta * cotbeta);
+          double clusterCharge_cut = clustChargeMaxPerLength_ * drdz;
+
+          float locBx = 1.;
+          if (cotbeta < 0.)
+            locBx = -1.;
+          float locBz = locBx;
+          if (cotalpha < 0.)
+            locBz = -locBx;
+
+          auto detId = detIdObj.rawId();
+          bool Isnew = false;
+          int DetId_index = -1;
+          for (int i = 0; i < (int)iHists.BPixnewDetIds_.size(); i++) {
+            if (detId == (unsigned int)iHists.BPixnewDetIds_[i])
+              Isnew = true;
+            DetId_index = i;
+          }
+
+          int TemplID = templateDBobject_->getTemplateID(detId);
+          templ.interpolate(TemplID, cotalpha, cotbeta, locBz, locBx);
+          qScale_ = templ.qscale();
+          rQmQt_ = templ.r_qMeas_qTrue();
+
+          // Surface deformation
+
+          const auto& lp_pair = surface_deformation(topol, tsos, recHitPix);
+
+          LocalPoint lp_track = lp_pair.first;
+          LocalPoint lp_rechit = lp_pair.second;
+
+          rechitCorr_.x = lp_rechit.x();
+          rechitCorr_.y = lp_rechit.y();
+          trackhitCorrX_ = lp_track.x();
+          trackhitCorrY_ = lp_track.y();
+
+          if (notInPCL_) {
+            SiPixelLorentzAngleTreeBarrel_->Fill();
+          }
+
+          // is one pixel in cluster a large pixel ? (hit will be excluded)
+          bool large_pix = false;
+          for (int j = 0; j < pixinfo_.npix; j++) {
+            int colpos = static_cast<int>(pixinfo_.col[j]);
+            if (pixinfo_.row[j] == 0 || pixinfo_.row[j] == 79 || pixinfo_.row[j] == 80 || pixinfo_.row[j] == 159 ||
```

Comment:
maybe this interface can be re-used?
https://github.com/cms-sw/cmssw/blob/c9e62f954fbc3f1dd5b5a9e9b562e3f5833bb9e6/Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h#L139

https://github.com/cms-sw/cmssw/blob/c9e62f954fbc3f1dd5b5a9e9b562e3f5833bb9e6/Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h#L141

Reply:
@tsusa please clarify why the previous usage.